### PR TITLE
Introduce new TypedData types

### LINKED ISSUE #297: Java runtime with Event hub, Binary datatype and Cardinality many
#### Investigative information

When posting a byte[] message to the event hub, the input message that I receive at Azure function is of type char[] and not byte[] and I get errors. Changing the Cardinality as "One", I get proper byte[], but it is not what cx needs.

Example:
@FunctionName("EventHubCall")
    public void eventHubProcessor(
            @EventHubTrigger(name = "msg", eventHubName = "%eventHubName%", 
            connection = "eventHubConnection", dataType = "binary", 
            cardinality = Cardinality.MANY) List<byte[]> messages,
            final ExecutionContext context)

#### Expected behavior

The messages.get(0) received back byte[]

#### Actual behavior

The messages.get(0) received back as char [] not byte[]

#### Known workarounds

N/A

#### Related information 

Provide any related information 

* Programming language used Java

### PR DESCRIPTION
**Pull request by amamounelsayed**
- Support CollectionSting, CollectionLong, CollectionDouble, CollectionByte[]
Fixes #297

### CODE REVIEW

File: src/main/java/com/microsoft/azure/functions/worker/handler/WorkerInitRequestHandler.java [hunk 1]
```diff
@@ -14,6 +14,7 @@ public WorkerInitRequestHandler() {
     @Override
     String execute(WorkerInitRequest request, WorkerInitResponse.Builder response) {
         response.setWorkerVersion(Application.version());
+        response.putCapabilities("TYPED_DATA_COLLECTION_SUPPORTED", "TRUE");
```
Comment (pragnagopa) [inline src/main/java/com/microsoft/azure/functions/worker/handler/WorkerInitRequestHandler.java]: Same as host: Change to pascal case `TypedDataCollection`

File: src/test/java/com/microsoft/azure/functions/worker/binding/tests/Utility.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+import sun.reflect.generics.reflectiveObjects.ParameterizedTypeImpl;
+
+import java.lang.reflect.Type;
+import java.util.List;
+
+public class Utility {
+    public static Type getActualType(Class clazz) {
+        return ParameterizedTypeImpl.make(List.class, new Type[]{clazz}, null);
+
```
Comment (pragnagopa) [inline src/test/java/com/microsoft/azure/functions/worker/binding/tests/Utility.java]: Nit: Delete empty line. 
Tests look good!

**Comment by pragnagopa** [comment]
@amamounelsayed - Please pull in the changes from proto file release: https://github.com/Azure/azure-functions-language-worker-protobuf/tree/v1.3.0-protofile

File: src/main/java/com/microsoft/azure/functions/worker/binding/RpcCollectionDoubleDataSource.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import sun.reflect.generics.reflectiveObjects.ParameterizedTypeImpl;
+
+import java.lang.reflect.Array;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.List;
+
+public final class RpcCollectionDoubleDataSource extends DataSource<List<Double>> {
+	public RpcCollectionDoubleDataSource(String name, TypedDataCollectionDouble value) {
+		super(name, value.getDoubleList(), COLLECTION_DATA_OPERATIONS);
```
[the remark this answers is not in the document]
Comment (pragnagopa) [reply src/main/java/com/microsoft/azure/functions/worker/binding/RpcCollectionDoubleDataSource.java]: You are right. Explicit conversions are needed to handle each collection type. For now, supporting just List is OK.

**Comment by pragnagopa** [comment]
Fixed proto file: https://github.com/Azure/azure-functions-language-worker-protobuf/releases/tag/v1.3.1-protofile

File: endtoendtests/Azure.Functions.Java.Tests.E2E/Azure.Functions.Java.Tests.E2E/Constants.cs [hunk 4]
```diff
@@ -37,10 +37,19 @@ public static class Constants
...
         public static string OutputEventHubQueueName = "test-eventhuboutput-java";
         public static string InputEventHubName = "test-input-java";
+
         public static string OutputJsonEventHubQueueName = "test-eventhuboutputjson-java";
         public static string InputJsonEventHubName = "test-inputjson-java";
+
         public static string OutputOneEventHubQueueName = "test-eventhuboutputone-java";
         public static string InputCardinalityOneEventHubName = "test-inputOne-java";
+
+        public static string OutputBinaryOneEventHubQueueName = "test-outputone-java-cardinality-one";
```
[the remark this answers is not in the document]
Comment (amamounelsayed) [reply endtoendtests/Azure.Functions.Java.Tests.E2E/Azure.Functions.Java.Tests.E2E/Constants.cs]: test-binary-output-java-cardinality-one

File: endtoendtests/src/main/java/com/microsoft/azure/functions/endtoend/EventHubTriggerTests.java [hunk 5]
```diff
@@ -77,6 +77,29 @@ public void TestEventHubOutputInputOne(
...
     }
 
+    @FunctionName("EventHubTriggerAndOutputBinaryCardinalityMany")
+    public void EventHubTriggerAndOutputBinaryCardinalityMany(
+            @EventHubTrigger(name = "messages", eventHubName = "test-inputbinary-java-cardinality-many", connection = "AzureWebJobsEventHubSender", dataType = "binary", cardinality = Cardinality.MANY) List<byte[]> messages,
+            @EventHubOutput(name = "output", eventHubName = "test-outputbinary-java-cardinality-many", connection = "AzureWebJobsEventHubSender") OutputBinding<byte[]> output,
+            final ExecutionContext context
+    ) {
+//        context.getLogger().info("Java Event Hub trigger received " + messages.size() +" messages");
+//        Byte[] test = messages.get(0);
```
[the remark this answers is not in the document]
Comment (amamounelsayed) [reply endtoendtests/src/main/java/com/microsoft/azure/functions/endtoend/EventHubTriggerTests.java]: It is in progress will be fixed in the next check in

**Comment by amamounelsayed** [comment]
@pragnagopa Based on the changed we did in the host we handle the list of string as collections. For example before when we have array of string we deal with is as list of Json. Now we will return it back as list of string and it is up to the user to handle the content in the way they like. referring to convertToStringArrayOrList.

File: endtoendtests/src/main/java/com/microsoft/azure/functions/endtoend/EventHubTriggerTests.java [hunk 6]
```diff
@@ -77,11 +77,40 @@ public void TestEventHubOutputInputOne(
         output.setValue(message);
     }
 
+    @FunctionName("EventHubTriggerAndOutputBinaryCardinalityManyListBinary")
+    public void EventHubTriggerAndOutputBinaryCardinalityManyListBinary(
+            @EventHubTrigger(name = "messages", eventHubName = "test-binary-input-cardinality-many-list-java", connection = "AzureWebJobsEventHubSender_2", dataType = "binary", cardinality = Cardinality.MANY) List<byte[]> messages,
+            @QueueOutput(name = "output", queueName = "test-binary-output-cardinality-many-list-java", connection = "AzureWebJobsStorage") OutputBinding<byte[]> output,
+            final ExecutionContext context
```
Comment (pragnagopa) [inline endtoendtests/src/main/java/com/microsoft/azure/functions/endtoend/EventHubTriggerTests.java]: Does this test work without using the new CLI ?

Comment (amamounelsayed) [reply endtoendtests/src/main/java/com/microsoft/azure/functions/endtoend/EventHubTriggerTests.java]: It is not but I commented the integration test, so this should not be called.

**Comment by pragnagopa** [comment]
Closing in favor of #316

**Pull request closed without merging**